stats/stats.go

```diff
@@ -71,6 +71,15 @@ type MetricType int
 
 // MarshalJSON serializes a MetricType as a human readable string.
 func (t MetricType) MarshalJSON() ([]byte, error) {
+	txt, err := t.MarshalText()
+	if err != nil {
+		return nil, err
+	}
+	return []byte(`"` + string(txt) + `"`), nil
```

Comment:
I have some worries that this might add additional allocations/work on the hot path of marshalling metric types, but I am not certain if this is used somewhere and whether it matters enough

Reply:
:man_shrugging: nothing seems to have been changed in the generated easyjson code in the cloud for marshaling, only for unmarshaling... but even if something changed, the previous code was terrible, so we should just get rid of JSON instead of trying these micro-optimizations...

Reply:
Although I probably should have just rewritten the whole `MetricType` and `ValueType` mess with enumer... the current code is just painful to look at, even after the fixes...

Reply:
I ran the `BenchmarkMetric*` tests from https://github.com/loadimpact/k6/blob/master/stats/cloud/bench_test.go, since they seem like realistic ones that should catch a performance regression here, and there doesn't seem to be one. I ran them with `-benchmem -count 5` and compared `master` (old) and this code (new), here's the benchstat output:
```
name                                           old time/op    new time/op    delta
MetricMarshal/10000-4                            14.9ms ± 2%    15.5ms ± 9%     ~     (p=0.151 n=5+5)
MetricMarshal/100000-4                            149ms ± 5%     150ms ± 3%     ~     (p=0.548 n=5+5)
MetricMarshal/500000-4                            738ms ± 5%     728ms ± 2%     ~     (p=0.548 n=5+5)
MetricMarshalWriter/10000-4                      14.8ms ± 1%    15.0ms ± 6%     ~     (p=0.690 n=5+5)
MetricMarshalWriter/100000-4                      148ms ± 3%     146ms ± 2%     ~     (p=0.151 n=5+5)
MetricMarshalWriter/500000-4                      703ms ± 2%     711ms ± 4%     ~     (p=0.690 n=5+5)
MetricMarshalGzip/10000_bestcompression-4         168ms ± 5%     167ms ± 6%     ~     (p=0.548 n=5+5)
MetricMarshalGzip/10000_default-4                71.5ms ± 1%    73.0ms ± 9%     ~     (p=0.841 n=5+5)
MetricMarshalGzip/10000_bestspeed-4              19.7ms ± 4%    19.0ms ± 2%   -3.24%  (p=0.032 n=5+5)
MetricMarshalGzip/100000_default-4                749ms ± 5%     748ms ± 3%     ~     (p=0.841 n=5+5)
MetricMarshalGzip/100000_bestspeed-4              203ms ± 5%     187ms ± 1%   -8.10%  (p=0.008 n=5+5)
MetricMarshalGzip/100000_bestcompression-4        1.78s ±28%     1.64s ± 3%     ~     (p=0.421 n=5+5)
MetricMarshalGzip/500000_bestcompression-4        8.38s ± 7%     8.06s ± 2%     ~     (p=0.690 n=5+5)
MetricMarshalGzip/500000_default-4                3.68s ± 5%     3.53s ± 8%     ~     (p=0.222 n=5+5)
MetricMarshalGzip/500000_bestspeed-4              945ms ± 4%     910ms ± 5%     ~     (p=0.056 n=5+5)
MetricMarshalGzipAll/10000_bestspeed-4           35.3ms ± 8%    34.6ms ± 5%     ~     (p=0.548 n=5+5)
MetricMarshalGzipAll/100000_bestspeed-4           348ms ± 3%     352ms ± 2%     ~     (p=0.421 n=5+5)
MetricMarshalGzipAll/500000_bestspeed-4           1.75s ± 6%     1.76s ± 2%     ~     (p=0.690 n=5+5)
MetricMarshalGzipAllWriter/10000_bestspeed-4     36.3ms ± 2%    36.6ms ± 4%     ~     (p=0.690 n=5+5)
MetricMarshalGzipAllWriter/100000_bestspeed-4     356ms ± 6%     355ms ± 4%     ~     (p=0.841 n=5+5)
MetricMarshalGzipAllWriter/500000_bestspeed-4     1.73s ± 5%     1.76s ± 1%     ~     (p=0.421 n=5+5)

name                                           old speed      new speed      delta
MetricMarshal/10000-4                           278MB/s ± 2%   268MB/s ± 8%     ~     (p=0.151 n=5+5)
MetricMarshal/100000-4                          277MB/s ± 4%   275MB/s ± 3%     ~     (p=0.548 n=5+5)
MetricMarshal/500000-4                          281MB/s ± 5%   285MB/s ± 2%     ~     (p=0.548 n=5+5)
MetricMarshalWriter/10000-4                     278MB/s ± 1%   275MB/s ± 5%     ~     (p=0.690 n=5+5)
MetricMarshalWriter/100000-4                    280MB/s ± 3%   284MB/s ± 2%     ~     (p=0.151 n=5+5)
MetricMarshalWriter/500000-4                    294MB/s ± 2%   291MB/s ± 4%     ~     (p=0.690 n=5+5)
MetricMarshalGzip/10000_bestcompression-4      24.5MB/s ± 4%  24.9MB/s ± 6%     ~     (p=0.548 n=5+5)
MetricMarshalGzip/10000_default-4              57.8MB/s ± 1%  56.8MB/s ± 8%     ~     (p=0.841 n=5+5)
MetricMarshalGzip/10000_bestspeed-4             210MB/s ± 4%   217MB/s ± 2%   +3.31%  (p=0.032 n=5+5)
MetricMarshalGzip/100000_default-4             55.3MB/s ± 4%  55.3MB/s ± 3%     ~     (p=0.841 n=5+5)
MetricMarshalGzip/100000_bestspeed-4            203MB/s ± 5%   221MB/s ± 1%   +8.76%  (p=0.008 n=5+5)
MetricMarshalGzip/100000_bestcompression-4     23.7MB/s ±23%  25.2MB/s ± 3%     ~     (p=0.421 n=5+5)
MetricMarshalGzip/500000_bestcompression-4     24.8MB/s ± 7%  25.7MB/s ± 2%     ~     (p=0.690 n=5+5)
MetricMarshalGzip/500000_default-4             56.3MB/s ± 5%  58.8MB/s ± 7%     ~     (p=0.222 n=5+5)
MetricMarshalGzip/500000_bestspeed-4            219MB/s ± 4%   228MB/s ± 5%     ~     (p=0.056 n=5+5)
MetricMarshalGzipAll/10000_bestspeed-4          117MB/s ± 8%   120MB/s ± 5%     ~     (p=0.548 n=5+5)
MetricMarshalGzipAll/100000_bestspeed-4         119MB/s ± 3%   117MB/s ± 3%     ~     (p=0.421 n=5+5)
MetricMarshalGzipAll/500000_bestspeed-4         118MB/s ± 6%   118MB/s ± 2%     ~     (p=0.690 n=5+5)
MetricMarshalGzipAllWriter/10000_bestspeed-4    114MB/s ± 2%   113MB/s ± 4%     ~     (p=0.690 n=5+5)
MetricMarshalGzipAllWriter/100000_bestspeed-4   116MB/s ± 6%   117MB/s ± 4%     ~     (p=0.841 n=5+5)
MetricMarshalGzipAllWriter/500000_bestspeed-4   120MB/s ± 5%   118MB/s ± 1%     ~     (p=0.421 n=5+5)

name                                           old alloc/op   new alloc/op   delta
MetricMarshal/10000-4                            4.38MB ± 1%    4.37MB ± 2%     ~     (p=1.000 n=5+5)
MetricMarshal/100000-4                           53.1MB ±12%    53.4MB ±11%     ~     (p=0.690 n=5+5)
MetricMarshal/500000-4                            311MB ± 0%     311MB ± 0%     ~     (p=0.825 n=5+5)
MetricMarshalWriter/10000-4                      1.24MB ±19%    1.37MB ±17%     ~     (p=0.548 n=5+5)
MetricMarshalWriter/100000-4                     20.1MB ±40%    22.1MB ±41%     ~     (p=0.548 n=5+5)
MetricMarshalWriter/500000-4                      104MB ± 0%     104MB ± 0%     ~     (p=0.333 n=4+5)
MetricMarshalGzip/10000_bestcompression-4         814kB ± 0%     814kB ± 0%     ~     (p=0.333 n=5+4)
MetricMarshalGzip/10000_default-4                 814kB ± 0%     814kB ± 0%     ~     (p=0.810 n=5+4)
MetricMarshalGzip/10000_bestspeed-4              1.21MB ± 0%    1.21MB ± 0%     ~     (p=0.476 n=5+5)
MetricMarshalGzip/100000_default-4                814kB ± 0%     814kB ± 0%     ~     (all equal)
MetricMarshalGzip/100000_bestspeed-4             1.21MB ± 0%    1.21MB ± 0%     ~     (all equal)
MetricMarshalGzip/100000_bestcompression-4        814kB ± 0%     814kB ± 0%     ~     (all equal)
MetricMarshalGzip/500000_bestcompression-4        814kB ± 0%     814kB ± 0%     ~     (all equal)
MetricMarshalGzip/500000_default-4                814kB ± 0%     814kB ± 0%     ~     (all equal)
MetricMarshalGzip/500000_bestspeed-4             1.21MB ± 0%    1.21MB ± 0%     ~     (all equal)
MetricMarshalGzipAll/10000_bestspeed-4           6.67MB ± 3%    6.55MB ± 3%     ~     (p=0.310 n=5+5)
MetricMarshalGzipAll/100000_bestspeed-4          64.8MB ± 0%    64.8MB ± 0%     ~     (p=0.286 n=5+4)
MetricMarshalGzipAll/500000_bestspeed-4           458MB ± 0%     458MB ± 0%     ~     (p=0.143 n=4+4)
MetricMarshalGzipAllWriter/10000_bestspeed-4     2.37MB ±18%    2.44MB ± 6%     ~     (p=0.690 n=5+5)
MetricMarshalGzipAllWriter/100000_bestspeed-4    35.6MB ±39%    43.9MB ±19%  +23.21%  (p=0.040 n=5+5)
MetricMarshalGzipAllWriter/500000_bestspeed-4     289MB ± 0%     289MB ± 0%     ~     (p=0.333 n=5+4)

name                                           old allocs/op  new allocs/op  delta
MetricMarshal/10000-4                               174 ± 1%       174 ± 1%     ~     (p=1.000 n=5+5)
MetricMarshal/100000-4                            1.67k ±11%     1.68k ±11%     ~     (p=0.730 n=5+5)
MetricMarshal/500000-4                            9.55k ± 0%     9.55k ± 0%     ~     (p=0.333 n=4+5)
MetricMarshalWriter/10000-4                         201 ± 3%       205 ± 4%     ~     (p=0.452 n=5+5)
MetricMarshalWriter/100000-4                      1.92k ±13%     1.98k ±14%     ~     (p=0.690 n=5+5)
MetricMarshalWriter/500000-4                      9.54k ± 0%     9.53k ± 0%     ~     (p=0.333 n=4+5)
MetricMarshalGzip/10000_bestcompression-4          16.0 ± 0%      16.0 ± 0%     ~     (all equal)
MetricMarshalGzip/10000_default-4                  16.0 ± 0%      16.0 ± 0%     ~     (all equal)
MetricMarshalGzip/10000_bestspeed-4                18.0 ± 0%      18.0 ± 0%     ~     (all equal)
MetricMarshalGzip/100000_default-4                 16.0 ± 0%      16.0 ± 0%     ~     (all equal)
MetricMarshalGzip/100000_bestspeed-4               18.0 ± 0%      18.0 ± 0%     ~     (all equal)
MetricMarshalGzip/100000_bestcompression-4         16.0 ± 0%      16.0 ± 0%     ~     (all equal)
MetricMarshalGzip/500000_bestcompression-4         16.0 ± 0%      16.0 ± 0%     ~     (all equal)
MetricMarshalGzip/500000_default-4                 16.0 ± 0%      16.0 ± 0%     ~     (all equal)
MetricMarshalGzip/500000_bestspeed-4               18.0 ± 0%      18.0 ± 0%     ~     (all equal)
MetricMarshalGzipAll/10000_bestspeed-4              206 ± 3%       202 ± 3%     ~     (p=0.325 n=5+5)
MetricMarshalGzipAll/100000_bestspeed-4           1.76k ± 0%     1.76k ± 0%   +0.20%  (p=0.048 n=5+4)
MetricMarshalGzipAll/500000_bestspeed-4           12.7k ± 0%     12.7k ± 0%   +0.13%  (p=0.029 n=4+4)
MetricMarshalGzipAllWriter/10000_bestspeed-4        217 ± 6%       220 ± 2%     ~     (p=0.595 n=5+5)
MetricMarshalGzipAllWriter/100000_bestspeed-4     2.18k ±19%     2.43k ±10%     ~     (p=0.206 n=5+5)
MetricMarshalGzipAllWriter/500000_bestspeed-4     12.7k ± 0%     12.7k ± 0%     ~     (p=0.333 n=5+4)

name                                           old ratio      new ratio      delta
MetricMarshalGzip/10000_bestcompression-4          15.5 ± 0%      15.5 ± 0%     ~     (all equal)
MetricMarshalGzip/10000_default-4                  13.8 ± 0%      13.8 ± 0%     ~     (p=0.444 n=5+5)
MetricMarshalGzip/10000_bestspeed-4                6.80 ± 0%      6.79 ± 0%     ~     (p=0.683 n=5+5)
MetricMarshalGzip/100000_default-4                 13.2 ± 0%      13.2 ± 0%     ~     (all equal)
MetricMarshalGzip/100000_bestspeed-4               6.79 ± 0%      6.80 ± 0%     ~     (p=0.238 n=5+4)
MetricMarshalGzip/100000_bestcompression-4         14.3 ± 0%      14.3 ± 0%     ~     (p=0.556 n=4+5)
MetricMarshalGzip/500000_bestcompression-4         14.3 ± 0%      14.2 ± 0%     ~     (p=0.095 n=4+5)
MetricMarshalGzip/500000_default-4                 13.2 ± 0%      13.2 ± 0%     ~     (p=0.333 n=5+4)
MetricMarshalGzip/500000_bestspeed-4               6.80 ± 0%      6.80 ± 0%     ~     (all equal)
```

everything (including allocations) seems the same, so I'm resolving this

Reply:
:+1: